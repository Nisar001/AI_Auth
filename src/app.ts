import 'reflect-metadata';
import express from 'express';
import cors from 'cors';
import dotenv from 'dotenv';
import { AppDataSource } from './config/db';
import { errorHandler, notFoundHandler } from './middlewares/errorHandler';
import { generalLimiter } from './middlewares/rateLimiter';
import authRoutes from './modules/routes/authRoutes';
import logger from './utils/logger';

// Load environment variables
dotenv.config();

const app = express();

// Security and parsing middleware
app.use(cors({
  origin: process.env.CORS_ORIGINS?.split(',') || ['http://localhost:3000'],
  credentials: true,
  methods: ['GET', 'POST', 'PUT', 'DELETE', 'PATCH'],
  allowedHeaders: ['Content-Type', 'Authorization']
}));

app.use(express.json({ limit: '10mb' }));
app.use(express.urlencoded({ extended: true, limit: '10mb' }));

// Apply rate limiting to all requests
app.use(generalLimiter);

// Health check endpoint
app.get('/health', (req, res) => {
  res.json({
    status: 'OK',
    timestamp: new Date().toISOString(),
    uptime: process.uptime(),
    environment: process.env.NODE_ENV || 'development'
  });
});

// API routes
app.use('/api/auth', authRoutes);

// Handle 404 errors
app.use(notFoundHandler);

// Global error handler
app.use(errorHandler);

// Initialize database connection
const initializeDatabase = async () => {
  try {
    await AppDataSource.initialize();
    logger.info('Database connected successfully');
  } catch (error) {
    logger.error('Database connection failed:', error);
    process.exit(1);
  }
};

// Graceful shutdown
process.on('SIGTERM', async () => {
  logger.info('SIGTERM received, shutting down gracefully');
  try {
    await AppDataSource.destroy();
    logger.info('Database connection closed');
    process.exit(0);
  } catch (error) {
    logger.error('Error during shutdown:', error);
    process.exit(1);
  }
});

export { app, initializeDatabase };