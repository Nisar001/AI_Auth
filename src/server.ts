import { app, initializeDatabase } from './app';
import logger from './utils/logger';

const PORT = parseInt(process.env.PORT || '3000', 10);

const startServer = async () => {
  try {
    // Initialize database connection
    await initializeDatabase();
    
    // Start the server
    const server = app.listen(PORT, () => {
      logger.info(`🚀 Server running on port ${PORT}`);
      logger.info(`📚 Environment: ${process.env.NODE_ENV || 'development'}`);
      logger.info(`🔗 Health check: http://localhost:${PORT}/health`);
      logger.info(`🔐 Auth API: http://localhost:${PORT}/api/auth`);
    });

    // Handle server errors
    server.on('error', (error: any) => {
      if (error.code === 'EADDRINUSE') {
        logger.error(`Port ${PORT} is already in use`);
      } else {
        logger.error('Server error:', error);
      }
      process.exit(1);
    });

    // Graceful shutdown on SIGINT
    process.on('SIGINT', () => {
      logger.info('SIGINT received, shutting down gracefully');
      server.close(() => {
        logger.info('Server closed');
        process.exit(0);
      });
    });

  } catch (error) {
    logger.error('Failed to start server:', error);
    process.exit(1);
  }
};

// Start the server
startServer();