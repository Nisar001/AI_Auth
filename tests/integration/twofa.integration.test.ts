// Mock Twilio and Nodemailer before any imports
jest.mock('twilio', () => {
  return jest.fn().mockImplementation(() => ({
    messages: { create: jest.fn().mockResolvedValue({ sid: 'mock-sid' }) }
  }));
});

jest.mock('nodemailer', () => {
  return {
    createTransport: jest.fn(() => ({
      sendMail: jest.fn().mockResolvedValue({ messageId: 'mock-message-id' }),
      verify: jest.fn().mockResolvedValue(true)
    }))
  };
});

jest.mock('../../src/utils/jwtUtils', () => {
  return {
    JwtUtils: {
      getTokenFromHeader: jest.fn(() => 'mock-access-token'),
      verifyAccessToken: jest.fn(() => ({
        userId: 'test-user-id',
        email: 'test@example.com',
      })),
    },
  };
});

jest.mock('../../src/middlewares/auth', () => {
  return {
    authenticateToken: jest.fn((req, res, next) => {
      req.user = {
        id: 'test-user-id',
        email: 'test@example.com',
        isEmailVerified: true,
        isPhoneVerified: true,
      };
      next();
    }),
    requireEmailVerification: jest.fn((req, res, next) => {
      if (!req.user?.isEmailVerified) {
        res.status(403).json({ error: 'Email verification required' });
        return;
      }
      next();
    }),
    requirePhoneVerification: jest.fn((req, res, next) => {
      if (!req.user?.isPhoneVerified) {
        res.status(403).json({ error: 'Phone verification required' });
        return;
      }
      next();
    }),
  };
});

jest.mock('../../src/middlewares/validation', () => {
  return {
    validateRequest: jest.fn((schema) => (req: any, res: any, next: any) => {
      next();
    }),
  };
});

jest.mock('../../src/utils/passwordUtils', () => {
  return {
    PasswordUtils: {
      comparePassword: jest.fn((providedPassword, storedPassword) => {
        return providedPassword === 'valid-password' && storedPassword === 'hashed-password';
      }),
    },
  };
});

jest.mock('../../src/services/otpService', () => {
  return {
    OtpService: jest.fn().mockImplementation(() => {
      return {
        generateAndSendOtp: jest.fn(async (user, method, context) => {
          return { otpId: 'mock-otp-id' };
        }),
        cleanupExpiredOtps: jest.fn(async () => {
          return true;
        }),
        generateSecretFor2FA: jest.fn(async (userId) => {
          return { secret: 'mock-secret', qrCode: 'mock-qr-code' };
        }),
      };
    }),
  };
});

import request from 'supertest';
import { app } from '../../src/app';
import dotenv from 'dotenv';

dotenv.config({ path: '.env.test' });

jest.mock('../../src/config/db', () => {
  return {
    AppDataSource: {
      getRepository: jest.fn(() => {
        return {
          findOne: jest.fn(async (query: any) => {
            if (query.where.id === 'test-user-id') {
              return {
                id: 'test-user-id',
                email: 'test@example.com',
                isEmailVerified: true,
                isPhoneVerified: true,
                password: 'hashed-password',
              };
            }
            return null;
          }),
          save: jest.fn(async (entity: any) => {
            return { id: 'test-user-id', ...entity };
          }),
        };
      }),
    },
  };
});

// Reset mocks before each test
beforeEach(() => {
  jest.clearAllMocks();
});

describe('2FA Integration', () => {
  let accessToken: string;

  beforeEach(() => {
    accessToken = 'mock-access-token';
  });

  // Refactor the 2FA setup test
  it('should initiate 2FA setup', async () => {
    const res = await request(app)
      .post('/api/auth/setup-2fa')
      .set('Authorization', `Bearer ${accessToken}`)
      .send({ method: 'email', password: 'valid-password' });

    expect(res.status).toBe(200); // Ensure successful initiation
    expect(res.body).toMatchSnapshot({
      data: {
        instructions: expect.any(Array),
        method: 'email',
        status: 'setup_initiated',
      },
      message: expect.any(String),
      statusCode: expect.any(Number),
      success: expect.any(Boolean),
    });
  });
});
