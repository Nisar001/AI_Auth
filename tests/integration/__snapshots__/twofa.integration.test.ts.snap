// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`2FA Integration should initiate 2FA setup 1`] = `
{
  "data": {
    "instructions": Any<Array>,
    "method": "email",
    "status": "setup_initiated",
  },
  "message": Any<String>,
  "statusCode": Any<Number>,
  "success": Any<Boolean>,
}
`;
